keep/action/action.py

```diff
@@ -81,13 +51,80 @@ def _check_throttling(self, action_name):
         return throttle.check_throttling(action_name, alert_id)
 
     def _run_foreach(self):
-        foreach_iterator = self.context_manager.get_actionable_results()
-        for val in foreach_iterator:
-            self.context_manager.set_for_each_context(val)
-            rendered_value = self.io_handler.render_context(self.provider_context)
-            self.provider.notify(**rendered_value)
+        # the item holds the value we are going to iterate over
+        items = self.io_handler.render(self.config.get("foreach"))
+        # apply ALL conditions (the decision whether to run or not is made in the end)
+        for item in items:
+            self.context_manager.set_for_each_context(item)
+            self._run_single()
 
     def _run_single(self):
+        # First, execute all conditions
+
+        # Initialize all conditions
+        conditions = [
+            ConditionFactory.get_condition(
+                condition.get("type"),
+                condition.get("name", None) or condition.get("type"),
+                condition,
+            )
+            for condition in self.conditions
```

Comment:
I don't sure I understand, but since condition have state, we can't re-initialize all conditions when we work in `foreach` context (because the condition applies again and again per input)

Reply:
self.conditions = [ConditionFactory.get_condition(
                condition.get("type"),
                condition.get("name", None) or condition.get("type"),
                condition,
            ) for condition in self.config.get("condition", []) or something

wouldn't that work?

Reply:
no, because then when you'll apply condition (condition.apply) it will be "dirty". condition should work once only.

Reply:
until decided other (and then will need to change the class)